Utils/UpdateList.cs

```diff
@@ -16,36 +15,57 @@ public class UpdateList
 		private readonly List<int> deadFunctionIndices = new List<int>();
 
 		private bool isUpdating;
-		private int currentUpdatingIndex;
 
+		/// <summary>
+		/// Returns true if the update list contains this function
+		/// </summary>
+		/// <param name="updateFunction">The update function to test for the presence of</param>
+		/// <returns>True if the function is in the update list, false if not</returns>
+		public bool Contains(Action<float> updateFunction)
+		{
+			var index = updateFunctions.IndexOf(updateFunction);
+			return index != -1 && !deadFunctionIndices.Contains(index);
+		}
+
+		/// <summary>
+		/// Adds the given function to the update list. It must not already be added.
+		/// </summary>
+		/// <param name="updateFunction">The update function to add to the update list</param>
+		/// <exception cref="Exception">Throws an Exception if the function was already added</exception>
 		public void Add(Action<float> updateFunction)
 		{
 			var index = updateFunctions.IndexOf(updateFunction);
 			if (index >= 0)
 			{
-				// Someone wants to add it back right after removed it
-				// Possibally, Abort() then Play() in an update callback
+				// if the function exists but it's marked as dead we should, unmark it (to re add it)
 				if (deadFunctionIndices.Contains(index))
 				{
-					// We just cancel the removement
 					deadFunctionIndices.Remove(index);
 					return;
 				}
-				Debug.LogWarning("You should not add the same update function more than once into update container.");
-				return;
+
+				throw new Exception("Cannot add the given function. It is already in the list");
 			}
 
 			updateFunctions.Add(updateFunction);
 		}
 
+		/// <summary>
+		/// Removes the given update function from the update list
+		/// </summary>
+		/// <param name="updateFunction">The update function to remove from the list</param>
+		/// <exception cref="Exception">Throws if the given update function was not in the list</exception>
 		public void Remove(Action<float> updateFunction)
 		{
 			var index = updateFunctions.IndexOf(updateFunction);
-			if (index < 0) return;
+			if (index < 0)
+			{
+				throw new Exception("Cannot remove the given function. It was not found in the list");
+			}
 
-			// if we are currently updating (and not updating the current one) we mark it as dead,
+			// if we are currently updating we mark it as dead,
 			// otherwise (else) we can remove it straight away
-			if (isUpdating && currentUpdatingIndex != index)
+			if (isUpdating)
```

Comment:
I can't comment on line 77 just below (thanks github), but couldn't the instant removal in the else still break it? We're not currently updating, but it still shifts the indexes down one, so the wrong index could be then removed in the removal loop right? Apologies if I'm repeating something we already talked about

Reply:
deadIndices will only contain indices during the update loop, it should be cleared the end of the frame, (the fact that it wasn't cleared before was the problem), meaning outside of the update we are free to manipulate the list outside of the update loop.